packages/labs/ssr-dom-shim/src/lib/events.ts

```diff
@@ -0,0 +1,537 @@
+
+// Shim the global EventTarget object
+const EventTargetShim = class EventTarget implements EventTargetInterface {
+  private __eventListeners = new Map<
+    string,
+    Map<EventListenerOrEventListenerObject, AddEventListenerOptions>
+  >();
+  private __captureEventListeners = new Map<
+    string,
+    Map<EventListenerOrEventListenerObject, AddEventListenerOptions>
+  >();
+  __eventTargetParent: EventTarget | undefined;
+  __host: EventTarget | undefined;
+
+  addEventListener(
+    type: string,
+    callback: EventListenerOrEventListenerObject | null,
+    options?: AddEventListenerOptions | boolean
+  ): void {
+    if (!callback) {
+      return;
+    }
+    const eventListenersMap = isCaptureEventListener(options)
+      ? this.__captureEventListeners
+      : this.__eventListeners;
+    let eventListeners = eventListenersMap.get(type);
+    if (!eventListeners) {
+      eventListeners = new Map();
+      eventListenersMap.set(type, eventListeners);
+    } else if (eventListeners.has(callback)) {
+      return;
+    }
+
+    const normalizedOptions =
+      typeof options === 'object' && options ? options : {};
+    normalizedOptions.signal?.addEventListener('abort', () =>
+      this.removeEventListener(type, callback, options)
+    );
+    eventListeners.set(callback, normalizedOptions ?? {});
+  }
+  removeEventListener(
+    type: string,
+    callback: EventListenerOrEventListenerObject | null,
+    options?: EventListenerOptions | boolean
+  ): void {
+    if (!callback) {
+      return;
+    }
+    const eventListenersMap = isCaptureEventListener(options)
+      ? this.__captureEventListeners
+      : this.__eventListeners;
+    const eventListeners = eventListenersMap.get(type);
+    if (eventListeners) {
+      eventListeners.delete(callback);
+      if (!eventListeners.size) {
+        eventListenersMap.delete(type);
+      }
+    }
+  }
+  dispatchEvent(event: Event): boolean {
+    const composedPath: EventTarget[] = [this];
+    let parent = this.__eventTargetParent;
+    if (event.composed) {
+      while (parent) {
+        composedPath.push(parent);
+        parent = parent.__eventTargetParent;
+      }
+    } else {
+      // If the event is not composed and the event was dispatched inside
+      // shadow DOM, we need to stop before the host of the shadow DOM.
+      while (parent && parent !== this.__host) {
+        composedPath.push(parent);
+        parent = parent.__eventTargetParent;
+      }
+    }
+
+    // We need to patch various properties that would either be empty or wrong
+    // in this scenario.
+    let stopPropagation = false;
+    let stopImmediatePropagation = false;
+    let eventPhase = 0;
+    let currentTarget: EventTarget | null = null;
+    const originalStopPropagation = event.stopPropagation;
+    const originalStopImmediatePropagation = event.stopImmediatePropagation;
+    Object.defineProperties(event, {
+      target: {
+        value: this,
```

Comment:
Does this need to do event re-targeting? See https://dom.spec.whatwg.org/#retarget

Reply:
I was not sure whether this is necessary for our limited use case.
But I could probably adapt that. My interpretation is that I could track a target similar to eventPhase and currentTarget. To be spec compliant I would probably also have to calculate eventPhase dependent on that.

Reply:
Yeah, it's hard to say. We also don't have a shadowRoot for the composedPath, so that's another infidelity, right?

I guess I'd be inclined to add retargetting if it is simple since the actual target setting is almost as straightforward as currentTarget (set target to the element if it has a shadowRoot, which I think we know?), and yes that would be in both phases.

I'm ok skipping this and documenting the limitation if it becomes too hard to manage.

Reply:
I'll have a look at what can be done. 👍

Reply:
Was a bit more complicated than I thought, but it should retarget correctly.